File: packages/examples/todo/backends/typescript/src/commands/delete.ts
/**
 * @fileoverview todo.delete command
 */

import { z } from 'zod';
import { defineCommand, success, failure } from '@afd/server';
import { store } from '../store/memory.js';

const inputSchema = z.object({
	id: z.string().min(1, 'Todo ID is required'),
});

export interface DeleteResult {
	deleted: boolean;
	id: string;
}

export const deleteTodo = defineCommand<typeof inputSchema, DeleteResult>({
	name: 'todo-delete',
	description: 'Delete a todo item',
	category: 'todo',
	mutation: true,
	version: '1.0.0',
	input: inputSchema,
	errors: ['NOT_FOUND'],

	async handler(input) {
		const existing = store.get(input.id);
		if (!existing) {
			return failure({
				code: 'NOT_FOUND',
				message: `Todo with ID "${input.id}" not found`,
				suggestion: 'Use todo.list to see available todos',
			});
		}

		const deleted = store.delete(input.id);

		if (!deleted) {
			return failure({
				code: 'NOT_FOUND',
				message: `Todo with ID "${input.id}" not found`,
				suggestion: 'Use todo.list to see available todos',
			});
		}

		return success(
			{ deleted: true, id: input.id },
			{
				reasoning: `Deleted todo "${existing.title}"`,
				confidence: 1.0,
				warnings: [
					{
						code: 'PERMANENT',
						message: 'This action cannot be undone',
						severity: 'info',
					},
				],
			}
		);
	},
});

File: packages/examples/todo/backends/rust/src/commands/delete.rs
use super::{not_found, ok, schema, Command};
use crate::store::TodoStore;
use afd::{failure, CommandResult};
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use std::sync::Arc;

pub struct Delete(pub Arc<TodoStore>);

#[derive(Deserialize)]
pub struct Input {
    id: String,
}

#[derive(Serialize)]
struct Output {
    deleted: bool,
    id: String,
}

impl Command for Delete {
    type Input = Input;
    const NAME: &'static str = "todo-delete";
    const DESCRIPTION: &'static str = "Delete a todo";
    const MUTATION: bool = true;

    fn schema() -> Value {
        json!({
            "type": "object",
            "properties": { "id": schema::id() },
            "required": ["id"]
        })
    }

    fn run(&self, input: Input) -> CommandResult<Value> {
        match self.0.delete(&input.id) {
            Some(todo) => {
                let reasoning = format!("Deleted todo \"{}\"", todo.title);
                let output = Output {
                    deleted: true,
                    id: input.id,
                };
                ok(&output, reasoning, 1.0, Vec::new())
            }
            None => failure(not_found(&input.id)),
        }
    }
}
